File: javascript/ocr-numbers/ocr-numbers.js
// @ts-check

const { freeze: icy } = Object, { from } = Array, SPC = / +$|\n/g,

      clean = (s='') => s.replace(SPC, ''), digits = (s='') => s.length/12 | 0,
      digitIdx = (num='', idx=0) => ~(idx = DIGITS.indexOf(num)) ? idx : '?',

      multiScan = (ocr='') => icy(ocr.split(REGEX).map((s, _, __, c=clean(s)) =>
        icy(from({ length: digits(s) }, (_, i) => scanDigit(i, c))))),

      REGEX = /(?<=\n\s{3,})\n/, OCR = ' _     _  _     _  _  _  _  _ '
        + '| |  | _| _||_||_ |_   ||_||_|' + '|_|  ||_  _|  | _||_|  ||_| _|',

      DIGITS = icy(from({ length: 10 }, (_, i) => scanDigit(i)));

function scanDigit(pos=0, ocr=OCR) {
  const len = ocr.length, cols = len/3 | 0, idx = pos*3;
  for (var digit='', i = idx; i < len; i += cols) digit += ocr.substr(i, 3);
  return digit; }

export function convert(ocr=OCR, blocks = multiScan(ocr)) {
  for (var val = '', i = 0; i < blocks.length; val += ',')
    for (const digit of blocks[i++]) val += digitIdx(digit);
  return console.log(ocr + '\n', val = val.slice(0, -1)), val; }

File: python/ocr-numbers/ocr_numbers.py
OCR = (' _     _  _     _  _  _  _  _ ', '| |  | _| _||_||_ |_   ||_||_|',
                                         '|_|  ||_  _|  | _||_|  ||_| _|')

ERR = 'Number of input %s is not a multiple of %s'
ROW_ERR, COL_ERR = ERR % ('lines', 'four'), ERR % ('columns', 'three')

TStr1d = tuple[str, str, str]; TStr2d = tuple[TStr1d, ...]

digitIdx = lambda num='': str(DIGITS.index(num)) if num in DIGITS else '?'

def clean(ocr: list[str]): return *(r for i, r in enumerate(ocr, 1) if i & 3),

def scanRows(ocr: TStr1d):
    return ''.join(digitIdx(scanDigit(i, ocr)) for i in size(ocr[0]))


def scanDigit(pos: int, ocr: TStr1d=OCR):
    r = slice(col := pos * 3, col + 3); return ''.join(row[r] for row in ocr)


size = lambda s='': range(len(s) // 3); DIGITS = *map(scanDigit, size(OCR[0])),

def convert(ocr: list[str]):
    if len(ocr) & 3: raise ValueError(ROW_ERR)
    if any(len(row) % 3 for row in ocr): raise ValueError(COL_ERR)

    iters = 3 * (iter(clean(ocr)),); multi: TStr2d = *zip(*iters),
    return print(*ocr, val := ','.join(map(scanRows, multi)), sep='\n') or val
